projects/packages/my-jetpack/src/products/class-module-product.php

```diff
@@ -131,4 +133,64 @@ public static function deactivate() {
 		return Jetpack::deactivate_module( static::$module_name );
 	}
 
+	/**
+	 * Get Products that support this product
+	 *
+	 * @return array
+	 */
+	public static function get_supporting_products() {
```

Comment:
> > What additional use cases does `get_supporting_products` cover when `get_supported_products` already exists?

> It's so hard to pick a right name 😅 
> 
>The existing `get_supported_products` is to get the product supported by the current product. For example: https://github.com/Automattic/jetpack/pull/31368/files#diff-96a3f7c8a02c743fc1ef8f8382674d897195688c05c45ef1d90b7422d0be465fR178
>
>The `get_supporting_products` function I added is on the other hand is to get what products support the current product if it makes sense.

Moving this discussion to a separate thread for better organization 🙂 

Two questions:
- Which products must specify a non-empty return value for this new function? I noticed we defined them for Jetpack Starter and Jetpack Security.
- What's the reason for defining this function instead of adding `stats` to Jetpack Complete's `get_supported_products`? Is it because Jetpack Complete is currently not defined as a product within My Jetpack?

I find it strange to redundantly define a parent-child relationship in two places: `get_supported_products` and `get_supporting_products`.

Take Jetpack Security, for example, which defines `array( 'backup', 'scan', 'anti-spam' )` as its "supported products". With this new `get_supporting_products`, wouldn't we also need to modify `backup`, `scan`, and `anti-spam` products to specify `array( $itself, 'jetpack_security' )` as the return value for their `get_supporting_products`?

If this is the case, then it'll be trivially easy to introduce regressions by updating the output of `get_supporting_products` of the child product but not the output of `get_supported_products` of the parent product.

Reply:
> Which products must specify a non-empty return value for this new function? I noticed we defined them for Jetpack Starter and Jetpack Security.

From the refactoring perspective, these Starter, Security and Stats are all Module Products. By extracting `has_required_plan` and `get_state_from_wpcom` to `Module_Product`, we managed to remove duplication and reduce API calls, but in the same time we'd need a method or member implemented in child classes to identify what products support the current products. Otherwise, `has_required_plan` wouldn't be able to be generalized.

A few of the products other than Module Products would better to be changed as well, but I want to limit the scope for the PR. And if it works well, we could go ahead and refactor those.

> What's the reason for defining this function instead of adding stats to Jetpack Complete's get_supported_products? Is it because Jetpack Complete is currently not defined as a product within My Jetpack?

I guess this is question about what is a product for My Jetpack. In my point of view, the products defined in My Jetpack are ones that we want to show Interstitial cards and also are "physical" products, meaning they are either modules, plugins or both. I don't think they are 'plans' or wpcom products.

It's bit mixed up for Security and Starter tho. I guess in the two cases, wpcom products become My Jetpack products. And that's the reason why `get_supported_products` is defined and used.

I'd like to loop in @Automattic/jetpack-agora for more information. Thanks team.

Reply:
@jsnmoon I think I'll revert the changes extracting `has_required_plan`, as I don't think it's actually working for Security or Starter. I'll only elevate `get_state_from_wpcom` instead.

Reply:
Sounds like a plan 👍